addons/controllers/clusterbootstrap_controller.go

```diff
@@ -1211,8 +1211,11 @@ func (r *ClusterBootstrapReconciler) GetDataValueSecretNameFromBootstrapPackage(
 			if !found {
 				// In this case, we expect the secretRef to be present under status subresource and its value gets updated by
 				// the corresponding controller. However, the config controller might not create the secret in time.
-				r.Log.Info("provider status does not have secretRef", "GVR", gvr)
-				return "", nil
+				// Note that as long as the providerRef is defined, there is no such case that the final state of the secretRef remain not shown.
+				// So we return a error for this to wait and reconcile.
+				err = fmt.Errorf("provider %s does not have secretRef on its status", provider.GetName())
```

Comment:
This would block the controller from installing other packages which are ready to go which is not what we want. 

My proposal is to skip creating package install resource on a cluster if providerRef is not nil and secret is empty with an info log. Since we watch the provider resource, any update to status with a secretRef will enqueue the controller to reconcile the cluster and result in package install creation.

Reply:
https://github.com/vmware-tanzu/tanzu-framework/blob/e4b205fcb028028f1b0df014e4bdf52b96f48245/addons/controllers/clusterbootstrap_controller.go#L1256-L1271

Hi Vijay, thanks for the comment. Based on my understanding, the ClusterBootstrap controller only watches events for tkr, ClusterBootstrap, and Secrets, but no provider resources. If we do not throw an error to enqueue again when providerRef is not nil and secret is empty, there might be a situation that the provider controller done creating the secret but not finished updating the secretRef on status, the secret creation event will trigger an ClusterBootstrap controller to enqueue but nothing happened due to the secretRef on provider's status is empty.

Unless we add all provider resources to ClusterBootstrap controller watch resources?

Reply:
Providers are watched separately since they are dynamic and can come and go based on new addon providers. See https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/clusterbootstrap_controller.go#L1084

Reply:
Hi @vijaykatam , thanks for pointing this out. I found that this dynamic watch is for additional packages only. Core packages providers like AntreaConfig, CalicoConfig, VSphereCPIConfig, KappControllerConfig, and VSphereCSIConfig are not included, am I right?
https://github.com/vmware-tanzu/tanzu-framework/blob/ada7554ccee6eb06f19592d57ef5a9472e50b67f/addons/controllers/clusterbootstrap_controller.go#L260-L276

Reply:
Good catch. I thought we were(perhaps a regression in our refactor). I am fixing it [here](https://github.com/vmware-tanzu/tanzu-framework/pull/3440) with tests.